receiver/signalfxreceiver/receiver.go

```diff
@@ -0,0 +1,255 @@
+		err = r.server.Close()
+	})
+	return err
+}
+
+func (r *sfxReceiver) handleReq(resp http.ResponseWriter, req *http.Request) {
+	// Tracing the request to make it visible via z-pages.
+	reqCtx := req.Context()
+	spanCtx, span := trace.StartSpan(reqCtx, r.config.Name())
+	defer span.End()
+
+	if req.Method != "POST" {
+		r.failRequest(resp, http.StatusBadRequest, responseInvalidMethod, nil, span)
+		return
+	}
+
+	if req.Header.Get("Content-Type") != "application/x-protobuf" {
+		r.failRequest(resp, http.StatusUnsupportedMediaType, responseInvalidContentType, nil, span)
+		return
+	}
+
+	encoding := req.Header.Get("Content-Encoding")
+	if encoding != "" && encoding != "gzip" {
+		r.failRequest(resp, http.StatusUnsupportedMediaType, responseInvalidEncoding, nil, span)
+		return
+	}
+
+	bodyReader := req.Body
+	if encoding == "gzip" {
+		var err error
+		bodyReader, err = gzip.NewReader(bodyReader)
+		if err != nil {
+			r.failRequest(resp, http.StatusBadRequest, responseErrGzipReader, err, span)
+			return
+		}
+	}
+
+	body, err := ioutil.ReadAll(bodyReader)
+	if err != nil {
+		r.failRequest(resp, http.StatusBadRequest, responseErrReadBody, err, span)
+		return
+	}
+
+	msg := &sfxpb.DataPointUploadMessage{}
+	if err := proto.Unmarshal(body, msg); err != nil {
+		r.failRequest(resp, http.StatusBadRequest, responseErrUnmarshalBody, err, span)
+		return
+	}
+
+	recvCtx := observability.ContextWithReceiverName(spanCtx, r.config.Name())
+	if len(msg.Datapoints) == 0 {
+		observability.RecordMetricsForMetricsReceiver(recvCtx, 0, 0)
+		return
+	}
+
+	md, numDroppedTimeseries := SignalFxV2ToMetricsData(r.logger, msg.Datapoints)
+
+	err = r.nextConsumer.ConsumeMetricsData(spanCtx, *md)
+	if err != nil {
+		observability.RecordMetricsForMetricsReceiver(
+			recvCtx,
+			len(msg.Datapoints),
+			len(msg.Datapoints))
+		r.failRequest(resp, http.StatusInternalServerError, responseErrNextConsumer, err, span)
+		return
+	}
+
+	observability.RecordMetricsForMetricsReceiver(
+		recvCtx,
+		len(msg.Datapoints),
+		numDroppedTimeseries)
+
+	resp.WriteHeader(http.StatusAccepted)
+}
+
+func (r *sfxReceiver) failRequest(
+	resp http.ResponseWriter,
+	httpStatusCode int,
+	msg string,
+	err error,
+	reqSpan *trace.Span,
+) {
+	resp.WriteHeader(httpStatusCode)
+	if msg != "" {
+		_, writeErr := resp.Write([]byte(msg))
+		if writeErr != nil {
+			r.logger.Warn(
+				"Error writing HTTP response message",
+				zap.Error(writeErr),
+				zap.String("receiver", r.config.Name()))
+		}
+	}
+
+	reqSpan.AddAttributes(
+		trace.Int64Attribute(conventions.AttributeHTTPStatusCode, int64(httpStatusCode)),
+		trace.StringAttribute(conventions.AttributeHTTPStatusText, msg))
+	traceStatus := trace.Status{
+		Code: trace.StatusCodeInvalidArgument,
+	}
+	if httpStatusCode == http.StatusInternalServerError {
+		traceStatus.Code = trace.StatusCodeInternal
+	}
+	if err != nil {
+		traceStatus.Message = err.Error()
+	}
+	reqSpan.SetStatus(traceStatus)
+
+	r.logger.Warn(
```

Comment:
Do we need this warning? This is business as usual. If someone sends us a lot of bad requests we will flood the logs. I think it will be better to keep internal metric/counter of bad requests instead and make this "Debug" level.

Reply:
I will move it to debug, regarding the metrics I am on the fence if I add a custom one now or if we do the work of having them on observability so they become standard/available to others.

Reply:
Sure, up to you on the metric, I don't have a strong opinion on this.

Reply:
Going with a custom metric now, will update when we have something uniform on core.

Reply:
Taking that back, I will add the metric late and will keep as Debug level for now.